cpp/scenario/gazebo/include/scenario/gazebo/Joint.h

```diff
@@ -0,0 +1,185 @@
+#include <ignition/gazebo/EntityComponentManager.hh>
+#include <ignition/gazebo/EventManager.hh>
+
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace scenario {
+    namespace base {
+        struct PID;
+        struct Limit;
+        enum class JointType
+        {
+            Invalid,
+            Fixed,
+            Revolute,
+            Prismatic,
+            Ball,
+        };
+        enum class JointControlMode
+        {
+            Idle,
+            Force,
+            Velocity,
+            Position,
+            PositionInterpolated,
+        };
+    } // namespace base
+    namespace gazebo {
+        class Joint;
+        class Model;
+    } // namespace gazebo
+} // namespace scenario
+
+class scenario::gazebo::Joint
+{
+public:
+    Joint();
+    virtual ~Joint();
+
+    // ============
+    // Gazebo Joint
+    // ============
+
+    uint64_t id() const;
+    bool initialize(const ignition::gazebo::Entity jointEntity,
+                    ignition::gazebo::EntityComponentManager* ecm,
+                    ignition::gazebo::EventManager* eventManager);
+    bool createECMResources();
+
+    bool historyOfAppliedJointForcesEnabled() const;
+    bool enableHistoryOfAppliedJointForces( //
+        const bool enable,
+        const size_t maxHistorySize = 100);
+    std::vector<double> getHistoryOfAppliedJointForces() const;
+
+    // ==========
+    // Joint Core
+    // ==========
+
+    size_t dofs() const;
+    std::string name() const;
+    base::JointType type() const;
+
+    base::Limit positionLimit() const;
+
+    base::JointControlMode controlMode() const;
+    bool setControlMode(const base::JointControlMode mode);
+
+    double controllerPeriod() const;
+
+    double maxEffort() const;
+    bool setMaxEffort(const double maxEffort);
+
+    base::PID pid() const;
+    bool setPID(const base::PID& pid);
+
+    // ==================
+    // Single DOF methods
+    // ==================
+
+    double force(const size_t dof = 0) const;
+    double position(const size_t dof = 0) const;
+    double velocity(const size_t dof = 0) const;
+
+    bool setForce(const double force, const size_t dof = 0);
+    bool setPositionTarget(const double position, const size_t dof = 0);
+    bool setVelocityTarget(const double velocity, const size_t dof = 0);
+    bool setAccelerationTarget(const double acceleration, const size_t dof = 0);
+
+    double positionTarget(const size_t dof = 0) const;
+    double velocityTarget(const size_t dof = 0) const;
+    double accelerationTarget(const size_t dof = 0) const;
+
+    bool resetPosition(const double position, const size_t dof = 0);
+    bool resetVelocity(const double velocity, const size_t dof = 0);
+    bool reset(const double position, //
+               const double velocity,
+               const size_t dof = 0);
+
+    // =================
+    // Multi DOF methods
+    // =================
+
+    std::vector<double> jointForce() const;
```

Comment:
same comment as https://github.com/robotology/gym-ignition/pull/158/files#r396516195

Reply:
Refer to https://github.com/robotology/gym-ignition/pull/158#discussion_r395493760